EIP 2604: Conditional EXTBALANCE Pricing (#2604)
ethereum/EIPs: EIPS/eip-draft-extbalance.md

---
eip: 2604
title: EXTBALANCE rename and conditional pricing
author: wjmelements.eth (@wjmelements)
discussions-to: https://ethereum-magicians.org/t/eip-2604-conditional-extbalance-repricing/
status: Draft
type: Standards Track
category: Core
created: 2020-04-16
requires: 1884
---

## Simple Summary

> Renames `BALANCE` to `EXTBALANCE` and reduces `EXTBALANCE` cost when querying the current contract.

Reviewer: ```suggestion
Renames `BALANCE (0x31)`mnemonic to `EXTBALANCE` and reduces `EXTBALANCE` cost when querying the current contract.
```

## Abstract
This EIP renames `BALANCE` to `EXTBALANCE`.
The gas cost of `EXTBALANCE` when the parameter matches the current address is reduced to a value slightly higher than the cost of `SELFBALANCE`.

## Motivation
As the primary purpose of `EXTBALANCE` following [EIP-1884](./eip-1884.md) is to query external accounts, the name is updated accordingly.

`EXTBALANCE` existed prior to `SELFBALANCE` so it is used by several immutable contracts, including the Uniswap-V1 template, that predate the Istanbul hard fork.
These contracts use `EXTBALANCE` to query their own balance.

## Specification

At block `N`, the `EXTBALANCE` operation changes from 700 gas to a conditional cost:
- If the parameter matches the current call context address, the cost is `GasSlowStep`, currently `10`
- Else the gas cost is `700`

## Rationale
Matching the rationale of [EIP-1884](./eip-1884.md), when querying its own balance a contract does not request a trie operation.
`SELFBALANCE` is cached and already on-hand because it is necessary for efficient operation of other opcodes such as CALL and CREATE.

> However, the gas cost of `EXTBALANCE` for the current account should be higher than the cost of `SELFBALANCE` because of the necessary comparison of the parameter to the current address.

Reviewer 2: <

---
eip: 2604
title: EXTBALANCE rename and conditional pricing

> author: wjmelements.eth <wjmelements@gmail.com> (@wjmelements)

Reviewer 3: ```suggestion
author: wjmelements.eth (@wjmelements)
```

Has to be one or the other, and we strongly prefer at least one author has a github username so EIPs can auto-merge correctly.

discussions-to: https://ethereum-magicians.org/t/eip-2604-conditional-extbalance-repricing/
status: Draft
type: Standards Track
category: Core
created: 2020-04-16
requires: 1884
---

## Simple Summary
Renames `BALANCE` to `EXTBALANCE` and reduces `EXTBALANCE` cost when querying the current contract.

> 

Reviewer 3: ```suggestion
```
For consistency across this repository.

## Abstract
This EIP renames `BALANCE` to `EXTBALANCE`.
The gas cost of `EXTBALANCE` when the parameter matches the current address is reduced to a value slightly higher than the cost of `SELFBALANCE`.

> 

Reviewer 3: ```suggestion
```

## Motivation

> As the primary purpose of `EXTBALANCE` following EIP1884 is to query external accounts, the name is updated accordingly.

Reviewer 3: ```suggestion
As the primary purpose of `EXTBALANCE` following [EIP-1884](./eip-1884.md) is to query external accounts, the name is updated accordingly.
```

Reviewer 3: All EIP's should be referred to as EIP-### and at *least* the first reference to any given EIP in an EIP should include a relative path link to it.

`EXTBALANCE` existed prior to `SELFBALANCE` so it is used by several immutable contracts, including the Uniswap-V1 template, that predate the Istanbul hard fork.
These contracts use `EXTBALANCE` to query their own balance.

## Specification

> At block `N`, the `EXTBALANCE` operation changes from 700 gas to a conditional cost:

Reviewer 3: The way this type of thing is worded in most EIPs is along the lines of:
> As of FORK_BLOCK_NUMBER, the `EXTBALANCE` (`0x##`) opcode will cost:
> - 10 if ...
> - 700 otherwise
